nearswap/src/pool.rs

```diff
@@ -170,3 +169,178 @@ impl Pool {
+        
+        // add liquidity again
+        let (near_added, tokens_added, shares_minted) = pool.add_liquidity(&caller, 200, 400, 0);
+
+        assert!(near_added == expected_near, "liquidity added is incorrect");
+        assert!(tokens_added == expected_tokens, "liquidity added is incorrect");
+        assert!(shares_minted == expected_shares, "liquidity added is incorrect");
+
+        let (expected_near2, expected_tokens2, expected_shares2) = expected_added_liquidity(100, 100, &pool);
+        // add liquidity again with ratio 1:1(100:100)
+        let (near_added2, tokens_added2, shares_minted2) = pool.add_liquidity(&caller, 100, 100, 0);
+
+        assert!(near_added2 == expected_near2, "liquidity added is incorrect");
+        assert!(tokens_added2 == expected_tokens2, "liquidity added is incorrect");
+        assert!(shares_minted2 == expected_shares2, "liquidity added is incorrect");
+    }
+
+    #[test]
+    fn withdraw_liquidity_with_min() {
+        init_blockchain();
+
+        let caller = "account".to_string();
+        let mut pool: Pool = setup_pool();
+
+        pool.add_liquidity(&caller, 100, 200, 0);
+
+        let min_ynear = 50;
+        let min_tokens = 100;
+        // withdraw liquidity with min required shares
+        let (ynear, token) = pool.withdraw_liquidity(&caller, min_ynear, min_tokens, 50);
+
+        assert!(ynear >= min_ynear, "Incorrect liquidity withdrawn");
+        assert!(token >= min_tokens, "Incorrect liquidity withdrawn");
+    }
+
+    #[test]
+    #[should_panic(expected = r#"E6: redeeming (ynear=50, tokens=100), which is smaller than the required minimum"#)]
+    fn withdraw_liquidity_with_min_fail() {
+        init_blockchain();
+
+        let caller = "account".to_string();
+        let mut pool: Pool = setup_pool();
+
+        pool.add_liquidity(&caller, 100, 200, 0);
+
+        let min_ynear = 55;
+        let min_tokens = 100;
+        // withdraw liquidity with min required shares
+        pool.withdraw_liquidity(&caller, min_ynear, min_tokens, 50);
+    }
+
+    #[test]
+    fn withdraw_liquidity_pool() {
+        init_blockchain();
+
+        let caller = "account".to_string();
+        let mut pool: Pool = setup_pool();
+
+        pool.add_liquidity(&caller, 100, 200, 0);
+
+        let (near_before, tokens_before, shares_before) = (pool.ynear, pool.tokens, pool.total_shares);
+        let (expected_near, expected_tokens) = expected_withdraw(50, &pool);
+
+        pool.withdraw_liquidity(&caller, 0, 0, 50);
+
+        // withdraw shares
+        assert!(pool.ynear == near_before - expected_near, "liquidity removed is incorrect");
+        assert!(pool.tokens == tokens_before - expected_tokens, "liquidity removed is incorrect");
+        assert!(pool.total_shares == shares_before - 50, "liquidity removed is incorrect");
+    }
+
+    #[test]
+    #[should_panic(expected = r#"attempt to subtract with overflow"#)]
+    fn withdraw_liquidity_pool_fail_scenario_1() {
```

Comment:
please add a comment describing the scenario. For the test case below - please add the description as well.

Reply:
ok